Project Private Events

expjazz opened the pull request:
## The project has a complex invitation system, where a event creator can invite other users to attend to the event. If you are not invited, you can request to participate (you will join a sign_up list), and if the creator wants it he will allow it(you will receive a formal invitation).

##### - The operations are restricted, so only the creators can invite new participants and accept the sign_up requests.

##### - The invited user can-t see the sign-up list or change anything in the event.

##### - The orders of the events in the root are crescent (according to the Date).

##### - There is a page where the user can see the events he attended and the future events he is invited to.

##### - An user can receive an invitation, but that does not mean he will attend. To do it so, he needs to click on the attend button on the events page (restrict only for invited users).

##### - We did integration test to see if the creation of events is working and if the authentication system is working.

##### - There is also tests for the models and its respective associations.

##### - We did a very basic UI since it was not a requirement of the project.

![events](https://user-images.githubusercontent.com/61479648/83047435-c0a93880-a01e-11ea-9304-15ee6e57e68a.png)
![invite option](https://user-images.githubusercontent.com/61479648/83047441-c1da6580-a01e-11ea-8d59-d8bf9207d14e.png)
![screenshot2](https://user-images.githubusercontent.com/61479648/83047874-60ff5d00-a01f-11ea-8cd7-182dfafe16fa.png)



We look forward to suggestions on how we can improve our code.

frankopkusianwar committed:
add bootsrtap

frankopkusianwar committed:
set up for testing

frankopkusianwar committed:
add login and signup pages

frankopkusianwar committed:
fix UI design

expjazz committed:
added events model. added creation model.

expjazz committed:
added invitation model and associations

frankopkusianwar committed:
add attendance model

frankopkusianwar committed:
add more fields

frankopkusianwar committed:
added event controllers

expjazz committed:
added invitations association from views

expjazz committed:
added show view for events and index of all events

expjazz committed:
Added test for User model. Added test for Event model

expjazz committed:
Added integration tests for the authentication system and event creation

frankopkusianwar committed:
add signedup model, list of users to be invited

frankopkusianwar committed:
add proper list style

expjazz committed:
Added restrictions for invitations. added Signup option for user

expjazz committed:
added option to attend

expjazz committed:
Added testing for models methods

frankopkusianwar committed:
add attended events view

frankopkusianwar committed:
add lint files

frankopkusianwar committed:
resolve lint errors

expjazz committed:
Added readme file updated

Redvanisation reviewed (changes requested):
# Changes required  ♻️ 

### Hello team  🌟

**Nice work on this project, but the following changes need to be done before moving on ⬇️** 

* [x] It has functional `login` page (authentication WITHOUT password, just providing a user email) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/login_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)).

#### Event Model
* [x] scope for upcoming events ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/scopes_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)).

* [x] scope for previous events ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/scopes_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)).

#### UsersController
##### Create Action
* [x] uses #build method on association, i.e. `current_user.events.build(event_params)` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/build_method_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)).

##### Index action
* [x] lists past events, i.e. `@past_events = Event.previous_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)).
* [x] lists ucomping events, i.e. `@upcoming_events = Event.upcoming_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)).

#### Show action
* [ ] lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)).
* [ ] lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)).
* [ ] lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)).

#### Database

* [ ] the events table has column `creator_id` of type `integer`.
* [ ] an index for `creator_id` column is added.
* [ ] the events_attendation/events_users/(any other name for join table between users and events) table has column `event_id` of type `integer`.
* [ ] an index for `event_id` column is added.
* [ ] the events_attendation/events_users/(any other name for join table between users and events) table has column `user_id` of type `integer`.
* [ ] an index for `user_id` column is added.

### _After making the required changes please submit a new code review request_

### Good luck! 🍀

expjazz committed:
added creators id to events

expjazz committed:
modified models as required by TSE

frankopkusianwar committed:
added upcoming and current events

expjazz committed:
Added option to attend to events

expjazz commented:
## Live Demo

We made the necessary adjustments:

- We initiate a session with only the user email
- Users can invite other users.
- A user can be invited, but that does not mean that he will Attend. We added a feature for that.
- We used the build method on the 'events#create'
- We also implemented the scopes for past and future events;
#### - To show the attended events, we used the scope with the addition of a method for checking if the user really went in the event.
- Rspec tests created for model associations and validations.
- integration tests for authentication and events management included.
- All foreign keys have corresponding indexes. white_check_mark


#### -  The Join table between Users and Events has the name of Invitations. It has all the requested keys and indexes.

We are looking forward to suggestions on how to improve the performance of the project;

expjazz committed:
Added signed up methods

expjazz committed:
Updated test for models, associations and integration tests

frankopkusianwar committed:
fix all lint errors

frankopkusianwar committed:
update live demo link

rOluochKe reviewed (approved):
Hello Team!
### You have done good and for also making changes from the previous reviews :clap: Your project looks good and therefore, it has been approved. Congratulations :boom: :100:
### You can merge and submit a project completion form from the dashboard. Good luck with the next project.
- Happy coding!
